### PR TITLE
Introduced initContainers to wait for DB/MQ availability

### LINKED ISSUE #11: Pod scheduling priority: Helm release could be stuck in a restart loop due Mongo/Rabbit clusters glued with a "Pending" state
> Moved from https://github.com/StackStorm/st2enterprise-dockerfiles/issues/80

Sometimes when deploying StackStorm cluster in K8s (especially with resource pressure), entire Helm deployment is stuck in a dead dependency loop: st2 pods keep restarting, rescheduling, recreating due to no MQ/DB connection, while Mongo and Rabbit cluster can't start because of st2 pods reschedule spam.

> Note that our HA deployment with 3 nodes for each DB and MQ, 2 for each st2 service creates a cluster of minimum `30` Pods

The solution could be trying K8s Pod priority https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/ which is beta starting from K8s `v1.11`.

This way we can prioritize scheduling for MongoDB, RabbitMQ and etcd clusters before st2 services.

### PR DESCRIPTION
**Pull request by arms11**
Closes #11

Added initContainers for waiting on DB and MQ dependencies to be formed and connected before proceeding to initialize dependent containers of ST2 services.

### CODE REVIEW

**Comment by arms11** [comment]
Attempt to resolve #11 . Please review.

**Comment by armab** [comment]
Nice!
First side effect from the change is that st2 Pods are reporting `Init` state until DB/MQ connections are established. I think that's helpful from the user's perspective and signalizes that st2 Pods are waiting for something: 

```
NAME                                            READY   STATUS             RESTARTS   AGE
pod/test-job-st2-apikey-load-bpvcp              0/1     Init:1/3           0          6m46s
pod/test-mongodb-0                              1/1     Running            0          6m46s
pod/test-mongodb-1                              0/1     CrashLoopBackOff   2          21s
pod/test-rabbitmq-0                             0/1     CrashLoopBackOff   0          6m46s
pod/test-redis-node-0                           2/2     Running            0          6m46s
pod/test-redis-node-1                           2/2     Running            0          6m20s
pod/test-redis-node-2                           2/2     Running            2          6m9s
pod/test-st2actionrunner-7f68887485-dfwd5       0/1     Init:1/2           0          6m45s
pod/test-st2actionrunner-7f68887485-g8zqm       0/1     Init:1/2           0          6m45s
pod/test-st2actionrunner-7f68887485-kf7jl       0/1     Init:1/2           0          6m46s
pod/test-st2actionrunner-7f68887485-ml5vv       0/1     Init:1/2           0          6m45s
pod/test-st2actionrunner-7f68887485-qb5kq       0/1     Init:1/2           0          6m45s
pod/test-st2api-6cc98fb84f-fmxtm                0/1     Init:1/2           0          6m46s
pod/test-st2api-6cc98fb84f-rf9dv                0/1     Init:1/2           0          6m46s
pod/test-st2auth-694df655c5-4rbk2               0/1     Init:1/3           0          6m46s
pod/test-st2auth-694df655c5-ggx7p               0/1     Init:1/3           0          6m45s
pod/test-st2client-7b689b8cb9-zg9td             0/1     Init:1/3           0          6m45s
pod/test-st2garbagecollector-67bbb56d54-xsdkp   0/1     Init:1/2           0          6m46s
pod/test-st2notifier-5d94775646-96vgp           0/1     Init:1/2           0          6m46s
pod/test-st2notifier-5d94775646-wwbv2           0/1     Init:1/2           0          6m46s
pod/test-st2rulesengine-7d986597c9-gz44n        0/1     Init:1/2           0          6m46s
pod/test-st2rulesengine-7d986597c9-zrg5k        0/1     Init:1/2           0          6m46s
pod/test-st2scheduler-856ccc64bb-grkxt          0/1     Init:1/2           0          6m46s
pod/test-st2scheduler-856ccc64bb-rxmrg          0/1     Init:1/2           0          6m46s
pod/test-st2sensorcontainer-59cc5cff7-tzm8b     0/1     Init:1/2           0          6m46s
pod/test-st2stream-6464974955-76vm2             0/1     Init:1/2           0          6m46s
pod/test-st2stream-6464974955-jlqcq             0/1     Init:1/2           0          6m46s
pod/test-st2timersengine-5c96d79cbd-xn545       0/1     Init:1/2           0          6m46s
pod/test-st2web-5d5d948d5-5qc5f                 1/1     Running            0          6m46s
pod/test-st2web-5d5d948d5-wkfgz                 1/1     Running            0          6m46s
pod/test-st2workflowengine-66555df5b4-kmt44     0/1     Init:1/2           0          6m46s
pod/test-st2workflowengine-66555df5b4-shdx6     0/1     Init:1/2           0          6m46s
```

:+1: 

Any downsides or edge cases anyone can think of?

File: templates/deployments.yaml [hunk 1]
```diff
@@ -43,6 +43,24 @@ spec:
       {{- end }}
       # Sidecar container for generating .htpasswd with st2 username & password pair and sharing produced file with the main st2auth container
```
Comment (armab) [inline templates/deployments.yaml:44]: These comment needs to be moved closer to the initContainer with beforementioned htpasswd

File: templates/deployments.yaml [hunk 2]
```diff
@@ -43,6 +43,24 @@ spec:
...
       initContainers:
+      - name: wait-for-db
+        image: busybox:1.28
+        command:
+          - 'sh'
+          - '-c'
+          - >
+            until nc -z -w 2 {{ .Release.Name }}-mongodb-headless 27017 && echo mongodb ok;
+              do sleep 2;
+            done
```
Comment (armab) [inline templates/deployments.yaml:54]:
```suggestion
            until nc -z -w 2 {{ .Release.Name }}-mongodb-headless 27017 && echo mongodb ok; do
              echo 'Waiting for MongoDB connection ...'
              sleep 2;
            done
```

Worth including a bit of logging there for users.

Same for RabbitMQ.

File: templates/deployments.yaml [hunk 3]
```diff
@@ -43,6 +43,24 @@ spec:
       {{- end }}
       # Sidecar container for generating .htpasswd with st2 username & password pair and sharing produced file with the main st2auth container
       initContainers:
+      - name: wait-for-db
+        image: busybox:1.28
+        command:
+          - 'sh'
+          - '-c'
+          - >
+            until nc -z -w 2 {{ .Release.Name }}-mongodb-headless 27017 && echo mongodb ok;
```
Comment (armab) [inline templates/deployments.yaml:52]: We'll need to take the port number from the MongoDB chart. You can find some hints in st2.conf templating.

Same for RabbitMQ.

**Review by armab — CHANGES_REQUESTED** [review]
Good idea, thanks for the contribution :+1: 

Left first comments on top of my head that need addressing.

We'll definitely need a Changelog for this enhancement.
I also think we may omit `wait-for` scenario for `st2client` Pod so it can start without `Init`.

**Comment by arms11** [comment]
Thanks @armab for the feedback. I will make the requested changes.

File: templates/deployments.yaml (hunk 1 above)
Comment (arms11) [reply templates/deployments.yaml]: Done

File: templates/deployments.yaml (hunk 2 above)
Comment (arms11) [reply templates/deployments.yaml]: Done

File: templates/deployments.yaml (hunk 3 above)
Comment (arms11) [reply templates/deployments.yaml]: Done

File: templates/_helpers.tpl [hunk 4]
```diff
@@ -94,3 +94,36 @@ Create the name of the stackstorm-ha service account to use
 {{- end -}}
 {{- end -}}
     
+{{- define "init-containers-wait-for-db" -}}
+{{- if index .Values "mongodb" "enabled" }}
```
Comment (armab) [inline templates/_helpers.tpl:98]: Just thought about that edge case. Perfect to have it already codified here :+1:

File: CHANGELOG.md [hunk 5]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## In Development
+* Implemented initContainers to wait for DB/MQ to be available for St2 Pods (#178)
```
Comment (armab) [inline CHANGELOG.md:4]: FYI I moved this changelog under the `In Development` section.

**Review by armab — APPROVED** [review]
Thanks a lot for this stability improvement! :+1:
